Bug: Use consistent method of checking for presence of info in connec…

josephsnyder opened the pull request:
…tion

Switch to using the .get function for accessing the potential information from
the S3 mirror objects.  If the key is not there, it will gracefully return
None instead of failing with a KeyError

See comment here: https://github.com/spack/spack/pull/24244#issuecomment-981678855

josephsnyder committed:
Bug: Standardize connection object for S3 connections

Ensure that mirror's to_dict function returns a syaml_dict object for all code
paths.

Switch to using the .get function for accessing the potential information from
the S3 mirror objects.  If the key is not there, it will gracefully return
None instead of failing with a KeyError

Additionally, check that the connection object is a dictionary before trying
to "get" from it.

Fixes: #27652

haampie commented:
Can we also get rid of the s3-related options for non-s3 urls? https://github.com/spack/spack/issues/27652#issuecomment-980053050

josephsnyder commented:
@haampie, Thanks for the additional feedback.  I'll get on it

alecbcs commented:
Hey all! Are there any lasting blockers to merge this? I'd really like to test if this PR fixes the Autamus build cache which was broken by #24244. If we can merge this then after the daily Spack container is built and released I can start checking it against Autamus.

stephenmsachs commented:
Could you please also add the following patch to fix the output of `spack mirror list` for s3 backed mirrors?

```
diff --git a/lib/spack/spack/mirror.py b/lib/spack/spack/mirror.py
index 9a48d83ecb..3091c7a368 100644
--- a/lib/spack/spack/mirror.py
+++ b/lib/spack/spack/mirror.py
@@ -110,9 +110,9 @@ def display(self, max_len=0):
             _display_mirror_entry(max_len, self._name, self._fetch_url)
         else:
             _display_mirror_entry(
-                max_len, self._name, self._fetch_url, "fetch")
+                max_len, self._name, self.fetch_url, "fetch")
             _display_mirror_entry(
-                max_len, self._name, self._push_url, "push")
+                max_len, self._name, self.push_url, "push")
 
     def __str__(self):
         name = self._name
```

Without the patch:
```
[root@ip-172-31-7-191 spack]# spack mirror list
infospherespackmirrortest    CommentedMap([('url', 's3://infospherespackmirrortest'), ('access_pair', [None, None]), ('access_token', None), ('profile', None), ('endpoint_url', None)]) (fetch)
infospherespackmirrortest    CommentedMap([('url', 's3://infospherespackmirrortest'), ('access_pair', [None, None]), ('access_token', None), ('profile', None), ('endpoint_url', None)]) (push)
spack-public                 https://mirror.spack.io
```
With the patch:
```
[root@ip-172-31-7-191 spack]# spack mirror list
infospherespackmirrortest    s3://infospherespackmirrortest (fetch)
infospherespackmirrortest    s3://infospherespackmirrortest (push)
spack-public                 https://mirror.spack.io
```

josephsnyder committed:
Re-enable pretty print of mirror information

Per @stephenmsachs comment, put the proper data into the
_display_mirror_entry function for `spack mirror list` display.

haampie commented:
@opadron this needs your review before it can be merged, if you don't have time to review before next week, please remove yourself from the reviewer list so this pr can be backported to 0.17.1. @josephsnyder code coverage is very low, can you potentially push a test?

josephsnyder commented:
@haampie,  the majority of the code changes are mostly not unittest-able.  That function only really gets exercised on a real connection attempt to an S3 bucket.  Unless, of course, we wanted to pull that specific information gathering chunk into a new function, in which case we could find some way to send it a mirror and get the two object's that we'd send into botocore back from it.  Any thoughts @opadron?

I don't know if that should necessarily stop this bug fix from being merged, but that's not my decision.
I'll revert this most recent push as it didn't add anything to the result.

opadron commented:
@josephsnyder yes, that's generally been the policy for this kind of code: unit test + mock what you can, but it's not worth reimplementing or mocking all of S3.  I wish where we drew the line was clearer, though.

It would be nice if we had a suite of integration tests that include real interactions with external services like S3.  Then, we could potentially add coverage from those tests.

josephsnyder committed:
Separate information gathering and test

Add a test for the capturing of the new S3 information

Additional fixes to more places where a default of None would be
best.

josephsnyder commented:
@opadron and @haampie,

I found a few more issues with the update once I ran autamus' script on my actual development machine.  I also separated out the information gathering code into its own function with some mock data and asserts in the test where we mock up the `spack.utils.s3` capabilities .  I think this one will give green results for both tests and coverage.  PTAL again when there is time.

haampie commented:
@opadron can you review?

opadron commented on lib/spack/spack/util/s3.py:
```diff
@@ -27,6 +27,30 @@ def _parse_s3_endpoint_url(endpoint_url):
+def get_mirror_s3_connection_info(connection):
+    s3_connection = {}
+
+    s3_connection_is_dict = connection and isinstance(connection, dict)
```
Seems like `s3_connection_is_dict` would evaluate to `False` if `connection == {}`.  Is this intentional?

josephsnyder replied:
Yes.  I don't see why we'd want to even attempt a check inside an empty dictionary,`{}`.  Do you see a reason to do that?

opadron replied:
No, I wasn't considering that much context.  I was just going by your variable name.  That's why I wondered what the intent was.

josephsnyder replied:
Oh, I see.  That makes complete sense.  Do you think it's worth changing to `has_s3_connection_information` or something similar?

haampie commented:
Note that this PR did not fix the generated config:

```
$ spack -e . mirror add x file://$PWD/bc
$ cat spack.yaml 
spack:
  # add package specs to the `specs` list
  specs: []
  view: true
  mirrors:
    x:
      fetch:
        url: file:///tmp/tmp.ttzinNcXLQ/bc
        access_pair:
        - null
        - null
        access_token: null
        profile: null
        endpoint_url: null
      push:
        url: file:///tmp/tmp.ttzinNcXLQ/bc
        access_pair:
        - null
        - null
        access_token: null
        profile: null
        endpoint_url: null
```

I anybody working on improving the config ^ format? It should not list AWS-related things.

josephsnyder commented:
@haampie, the configuration file update was done as a separate merge request: #27701
